mobile-ui/src/main/java/org/sagebionetworks/research/mobile_ui/recorder/distance/DistanceRecorder.java

```diff
@@ -0,0 +1,279 @@
+        public long getTotalTime() {
+            return this.lastLocation != null && this.firstLocation != null ?
+                    this.lastLocation.getTime() - this.firstLocation.getTime() : 0;
+        }
+
+        /**
+         * Returns the average speed of the user during the recorders run in meters per second.
+         * @return the average speed of the user during the recorders run in meters per second.
+         */
+        public double getAverageSpeed() {
+            return this.totalDistance / (this.getTotalTime() * 1e-6);
+        }
+
+        /**
+         * Returns the first location the user was recorded at by the recorder.
+         * @return the first location the user was recorded at by the recorder.
+         */
+        public Location getFirstLocation() {
+            return this.firstLocation;
+        }
+
+        /**
+         * Returns the location the user finished their recording at.
+         * @return the location the user finished their recording at.
+         */
+        public Location getLastLocation() {
+            return this.lastLocation;
+        }
+    }
+
+    /**
+     * Adds information that is only relevant to the CurrentState of the recorder to DistanceInfo.
+     */
+    public static class CurrentDistanceInfo extends DistanceInfo {
+        private double currentSpeed;
+
+        public CurrentDistanceInfo() {
+            super();
+            this.currentSpeed = 0;
+        }
+
+        /**
+         * Returns the most recently measured location of the user.
+         * @return the most recently measured location of the user.
+         */
+        public Location getCurrentLocation() {
+            // When the recorder is running the last is the most recent location the user was measured at.
+            return this.getLastLocation();
+        }
+
+        @Override
+        public void update(Location nextLocation) {
+            if (this.getLastLocation() != null) {
+                long timeChangeMicros = nextLocation.getTime() - this.getLastLocation().getTime();
+                float distanceChangeMeters = this.getLastLocation().distanceTo(nextLocation);
+                this.currentSpeed = distanceChangeMeters / (timeChangeMicros * 1e-6);
+            }
+
+            super.update(nextLocation);
+        }
+
+        /**
+         * Returns the most recently measured speed of the user.
+         * @return the most recently measured speed of the user.
+         */
+        public double getCurrentSpeed() {
+            return this.currentSpeed;
+        }
+    }
+
+    /**
+     * Subscribes to the same Location stream as the DistanceRecorder and maintains a Summary of the recorder's
+     * recordings that will complete when the recorder is finished. This class should be used when the recorder's
+     * findings are only needed after the recorder is done recording.
+     */
+    protected static class DistanceSummarySubscriber implements SummarySubscriber<DistanceInfo, Location>,
```

Comment:
Is this different than doing something like `locationObservable.scan(distanceAccumulator).last()`?